File: frontend/src/api/types/journey.ts
import type { IItem } from './item';

export interface IItemRequest {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  Journey: null;
  JourneyID: number;
  WarehouseID: number;
  Item: IItem;
  ItemID: number;
  counts: number;
  RequestedBy: null;
  RequestedByID: number;
}

export enum JourneyStatus {
  LOADING = 'LOADING',
  ON_THE_WAY = 'ON_THE_WAY',
}

export interface IWarehouse {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  Name: string;
}

export interface IWaypoint {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  Route: null;
  RouteID: number;
  Warehouse: IWarehouse;
  WarehouseID: IWarehouse['ID'];
  Duration: number;
}

export interface IRoute {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  Name: string;
  Waypoints: IWaypoint[];
  interval: number;
}

export interface IJourney {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  Route: IRoute;
  RouteID: number;
  DepartureTime: string;
  Departed: boolean;
  ItemRequests: IItemRequest[];
  Place: number;
}

File: backend/internal/entity/journey.go
package entity

import (
	"gorm.io/gorm"
	"time"
)

type Journey struct {
	gorm.Model
	Route         *Route
	RouteID       uint `gorm:"not null"`
	DepartureTime time.Time
	ItemRequests  []*RequestedItems `gorm:"foreignkey:JourneyID"`
	Departed      bool
	Place         int // 0 = at central warehouse, 1 = at first station, 2 = second, ...
}

type RequestedItems struct {
	gorm.Model
	Journey       *Journey
	JourneyID     uint `gorm:"not null"`
	Item          *Item
	ItemID        uint `gorm:"not null"`
	Counts        uint `json:"counts"`
	RequestedBy   *User
	RequestedByID uint `gorm:"not null"`
	Warehouse     *Warehouse
	WarehouseID   uint `gorm:"not null"`
	Arrived       bool
}
